#include "poscardwidget.h"
#include "ui_poscardwidget.h"

PosCardWidget::PosCardWidget(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::PosCardWidget)
{
    ui->setupUi(this);

    setupUI();
    createConnections();
}

PosCardWidget::~PosCardWidget()
{
    delete ui;
}

void PosCardWidget::setupUI()
{
    // –°—Ç–∏–ª—ñ–∑—É—î–º–æ QFrame (—Ä–æ–±–∏–º–æ –∫–∞—Ä—Ç–∫—É –∫—Ä–∞—Å–∏–≤–æ—é)
    ui->frame->setStyleSheet(R"(
        QFrame#frame {
            background-color: white;
            border: 1px solid #dadce0;
            border-radius: 8px;
        }
        QFrame#frame:hover {
            border: 1px solid #1a73e8; /* –°–∏–Ω—è —Ä–∞–º–∫–∞ –ø—Ä–∏ –Ω–∞–≤–µ–¥–µ–Ω–Ω—ñ –º–∏—à–∫–æ—é */
        }
        QLabel {
            border: none; /* –©–æ–± –ª–µ–π–±–ª–∏ –Ω–µ —É—Å–ø–∞–¥–∫—É–≤–∞–ª–∏ —Ä–∞–º–∫—É –≤—ñ–¥ —Ñ—Ä–µ–π–º—É */
        }
    )");

    // –î–æ–¥–∞—î–º–æ –ø—ñ–¥–∫–∞–∑–∫—É –Ω–∞ –∫–Ω–æ–ø–∫—É
    ui->toolButton->setToolTip("–ö–æ–ø—ñ—é–≤–∞—Ç–∏ —ñ–Ω—Ñ–æ—Ä–º–∞—Ü—ñ—é –ø—Ä–æ —Ü—é –∫–∞—Å—É");
}

void PosCardWidget::createConnections()
{
    // –ü—ñ–¥–∫–ª—é—á–∞—î–º–æ –∫–ª—ñ–∫ –ø–æ –∫–Ω–æ–ø—Ü—ñ –¥–æ –Ω–∞—à–æ–≥–æ —Å–ª–æ—Ç–∞
    connect(ui->toolButton, &QToolButton::clicked, this, &PosCardWidget::onCopyClicked);
}

void PosCardWidget::setData(const QJsonObject &json, const QString &clientName, const QString &terminalId)
{
    // –ó–±–µ—Ä—ñ–≥–∞—î–º–æ –¥–∞–Ω—ñ —É –∫–ª–∞—Å—ñ
    m_posData = json;
    m_clientName = clientName;
    m_terminalId = terminalId;

    // –í–∏—Ç—è–≥—É—î–º–æ –¥–∞–Ω—ñ –∑ JSON –±–µ–∑–ø–µ—á–Ω–æ
    int posId = json["pos_id"].toInt();
    QString manufacturer = json["manufacturer"].toString();
    QString model = json["model"].toString();
    QString zn = json["factory_number"].toString();
    QString fn = json["tax_number"].toString();
    QString softVer = json["version"].toString();
    QString mukVer = json["muk_version"].toString();
    QString regDate = json["reg_date"].toString();

    // –ó–∞–ø–æ–≤–Ω—é—î–º–æ UI

    ui->labelNameModel->setText(QString("<img src=':/res/Images/RRO_icon.png' width='16' height='16' align='middle'> –ö–∞—Å–∞ ‚Ññ%1 %2-%3").arg(posId).arg(manufacturer, model));
    ui->labelZnFn->setText(QString("‚öôÔ∏è –ó–ù: %1    –§–ù: %2").arg(zn, fn));
    ui->labelSoftMuk->setText(QString("üíª –ü–û: %1 | –ú–£–ö: %2").arg(softVer, mukVer));

    if (!regDate.isEmpty()) {
        ui->labelRegData->setText(QString("üìÖ –†–µ—î—Å—Ç—Ä–∞—Ü—ñ—è: %1").arg(regDate));
    } else {
        ui->labelRegData->setText("üìÖ –†–µ—î—Å—Ç—Ä–∞—Ü—ñ—è: –ù–µ–º–∞—î –¥–∞–Ω–∏—Ö");
    }
}

void PosCardWidget::onCopyClicked()
{
    // –§–æ—Ä–º—É—î–º–æ —Ç–µ–∫—Å—Ç —Ç–æ—á–Ω–æ –∑–∞ –≤–∞—à–∏–º —à–∞–±–ª–æ–Ω–æ–º
    int posId = m_posData["pos_id"].toInt();
    QString manufacturer = m_posData["manufacturer"].toString();
    QString model = m_posData["model"].toString();
    QString zn = m_posData["factory_number"].toString();
    QString fn = m_posData["tax_number"].toString();
    QString softVer = m_posData["version"].toString();
    QString mukVer = m_posData["muk_version"].toString();
    QString regDate = m_posData["reg_date"].toString();

    QStringList lines;

    // –®–∞–ø–∫–∞ –ê–ó–°
    lines << QString("üè¢ %1, –ê–ó–° %2").arg(m_clientName, m_terminalId);

    // –î–∞–Ω—ñ –∫–∞—Å–∏
    lines << QString("üßæ –ö–∞—Å–∞ ‚Ññ%1 %2-%3").arg(posId).arg(manufacturer, model);
    lines << QString("   –ó–ù: %1").arg(zn);
    lines << QString("   –§–ù: %1").arg(fn);
    lines << QString("   üíª –ü–û: %1 | –ú–£–ö: %2").arg(softVer, mukVer);

    if (!regDate.isEmpty()) {
        lines << QString("   üìÖ –†–µ—î—Å—Ç—Ä–∞—Ü—ñ—è: %1").arg(regDate);
    }

    // –í—ñ–¥–ø—Ä–∞–≤–ª—è—î–º–æ –≤ –±—É—Ñ–µ—Ä –æ–±–º—ñ–Ω—É
    QClipboard *clipboard = QApplication::clipboard();
    clipboard->setText(lines.join("\n"));
}
